File: practice-scala-base/src/leetcode/MaximumProduct.scala
package leetcode

/**
 * @author zoe
 **/
object MaximumProduct {
  def main(args: Array[String]): Unit = {
    println(maximumProduct(Array(9,1,5,6,7,2)))
  }

  def maximumProduct(nums: Array[Int]): Int = {
    val sorted = nums.sorted
    val length = nums.length
    (sorted(0)*sorted(1)*sorted(length-1)).max(sorted(length-1)*sorted(length-2)*sorted(length-3))
  }

}

File: practice-leetcode/src/main/java/cn/wllsrx/zoe/leetcode/MaximumProduct.java
package cn.wllsrx.zoe.leetcode;

import java.util.Arrays;

/**
 * @author zoe
 * <p>
 * 三个数的最大乘积
 * <p>
 * 给定一个整型数组，在数组中找出由三个数组成的最大乘积，并输出这个乘积。
 * <p>
 * 示例 1:
 * <p>
 * 输入: [1,2,3]
 * 输出: 6
 * 示例 2:
 * <p>
 * 输入: [1,2,3,4]
 * 输出: 24
 * 注意:
 * <p>
 * 给定的整型数组长度范围是[3,104]，数组中所有的元素范围是[-1000, 1000]。
 * 输入的数组中任意三个数的乘积不会超出32位有符号整数的范围。
 **/
public class MaximumProduct {

    public static void main(String[] args) {
        System.out.println(maximumProduct(new int[]{1, 2, 3, 4}));
    }

    private static int maximumProduct(int[] nums) {
        Arrays.sort(nums);
        return Math.max(nums[0] * nums[1] * nums[nums.length - 1],
                nums[nums.length - 1] * nums[nums.length - 2] *
                        nums[nums.length - 3]);
    }
}
